contracts/JoepegAuctionHouse.sol

```diff
@@ -0,0 +1,960 @@
+            auction.currency,
+            _collection,
+            _tokenId,
+            auction.startPrice,
+            block.timestamp,
+            auction.endTime
+        );
+    }
+
+    /// @notice Place bid on a running English Auction
+    /// @param _collection address of ERC721 token
+    /// @param _tokenId token id of ERC721 token
+    /// @param _currency address of currency to bid
+    /// @param _amount amount of currency to bid
+    function placeEnglishAuctionBid(
+        address _collection,
+        uint256 _tokenId,
+        address _currency,
+        uint256 _amount
+    ) public nonReentrant {
+        IERC20(_currency).safeTransferFrom(msg.sender, address(this), _amount);
+        _placeEnglishAuctionBid(_collection, _tokenId, _currency, _amount);
+    }
+
+    /// @notice Place bid on a running English Auction using AVAX/WAVAX
+    /// @param _collection address of ERC721 token
+    /// @param _tokenId token id of ERC721 token
+    /// @param _wavaxAmount amount of WAVAX to bid
+    function placeEnglishAuctionBidWithAVAXAndWAVAX(
+        address _collection,
+        uint256 _tokenId,
+        uint256 _wavaxAmount
+    ) public payable nonReentrant {
+        if (msg.value > 0) {
+            // Wrap AVAX into WAVAX
+            IWAVAX(WAVAX).deposit{value: msg.value}();
+        }
+        if (_wavaxAmount > 0) {
+            IERC20(WAVAX).safeTransferFrom(
+                msg.sender,
+                address(this),
+                _wavaxAmount
+            );
+        }
+        _placeEnglishAuctionBid(
+            _collection,
+            _tokenId,
+            WAVAX,
+            msg.value + _wavaxAmount
+        );
+    }
+
+    /// @notice Settles an English Auction
+    /// @dev Note:
+    /// - Can be called by creator at any time (including before the auction's end time to accept the
+    ///   current latest bid)
+    /// - Can be called by anyone after the auction ends
+    /// - Transfers funds and fees appropriately to seller, royalty receiver, and protocol fee recipient
+    /// - Transfers ERC721 token to last highest bidder
+    /// @param _collection address of ERC721 token
+    /// @param _tokenId token id of ERC721 token
+    function settleEnglishAuction(address _collection, uint256 _tokenId)
+        public
+    {
+        EnglishAuction memory auction = englishAuctions[_collection][_tokenId];
+        if (auction.creator == address(0)) {
+            revert JoepegAuctionHouse__NoAuctionExists();
+        }
+        if (auction.lastBidPrice == 0) {
+            revert JoepegAuctionHouse__EnglishAuctionCannotSettleWithoutBid();
+        }
+        if (
+            msg.sender != auction.creator && block.timestamp < auction.endTime
+        ) {
+            revert JoepegAuctionHouse__EnglishAuctionOnlyCreatorCanSettleBeforeEndTime();
+        }
+
+        _clearEnglishAuction(_collection, _tokenId);
+
+        // Settle auction using latest bid
+        if (auction.currency == WAVAX) {
```

Comment:
hm thats a good question. maybe we could enforce caller is EOA? @LouisMeMyself do you have any thoughts here?

Reply:
another idea just in case of emergency is adding an `onlyOwner` function that allows owner to end an auction (i.e. return NFT to seller and bid to bidder)

Reply:
I just realized that the bidder could also be a contract that can't receive AVAX directly.
Maybe the cancel could have an option to send WAVAX instead of AVAX

Reply:
hm I see — that's probably why `JoepegExchange` also just sends out WAVAX instead of AVAX. i think i'll do the same to avoid complicating the contract code